src/main/java/com/paloaltonetworks/osc/api/PANManagerSecurityGroupApi.java

```diff
@@ -1,110 +1,228 @@
+        }
+    }
+
+    @Override
+    public void deleteSecurityGroup(String sgId) throws Exception {
+        String sgTag = securityGroupTag(sgId);
+        Set<String> ipsOnMgr = fetchMgrIpsByTag(sgTag);
+        try {
+            for (String ip : ipsOnMgr) {
+                LOG.info("Deleting address {}", ip);
+                deleteAddress(ip);
+            }
+        } catch (Exception e) {
+            LOG.error("Exception deleting security group " + sgId, e);
+            throw e;
+        }
+
+        deleteSGTag(sgId);
+        this.panClient.configCommit();
+    }
+
+    @Override
+    public List<? extends ManagerSecurityGroupElement> getSecurityGroupList() throws Exception {
+        return null;
+    }
+
+    @Override
+    public ManagerSecurityGroupElement getSecurityGroupById(String sgId) throws Exception {
+        return null;
+    }
+
+    @Override
+    public void close() {
+    }
+
+    private void doUpdateSecurityGroup(String sgId, String name, SecurityGroupMemberListElement memberList)
+            throws Exception {
+        LOG.info("Populating security group {} with members: {}", name, memberList.toString());
+        String sgTag = securityGroupTag(sgId);
+
+        addSGTagAndCheck(sgTag);
+
+        Set<String> ipsToSet = computeIps(memberList);
+        Set<String> ipsOnMgr = fetchMgrIpsByTag(sgTag);
+        Set<String> tmp = new HashSet<>(ipsOnMgr);
+        ipsOnMgr.removeAll(ipsToSet);
+        ipsToSet.removeAll(tmp);
+
+        for (String ip : ipsToSet) {
+            LOG.info("Creating address {}", ip);
+            addAddress(ip);
+            LOG.info("Adding address {} to security group {} ", ip, name);
+            this.panClient.addTagToAddress(ip, sgTag, this.devGroup);
+        }
+
+        for (String ip : ipsOnMgr) {
+            this.panClient.removeTagFromAddress(ip, sgTag, this.devGroup);
+            LOG.info("Deleting address {}", ip);
+            deleteAddress(ip);
+        }
+
+        this.panClient.configCommit();
+    }
+
+    private Set<String> fetchMgrIpsByTag(String sgTag) throws Exception {
+        return this.panClient
+                .fetchAddressesWithTag(sgTag, this.devGroup).stream().map(ae -> ae.getName())
+                .collect(Collectors.toSet());
+    }
+
+    private Set<String> computeIps(SecurityGroupMemberListElement memberList) {
+        Set<String> rightIps = new HashSet<>();
+
+        for (SecurityGroupMemberElement member : memberList.getMembers()) {
+            rightIps.addAll(member.getIpAddresses());
+        }
+
+        return rightIps;
+    }
+
+    private void addSGTagAndCheck(String sgTag) throws Exception {
+        try {
+            LOG.info("Adding tag {} under xpath {}", sgTag, this.tagXpath);
+            addSGTag(sgTag); // OK if exists
+        } catch (Exception e) {
+            LOG.error("Exception adding security group " + sgTag, e);
+            throw e;
+        }
+
+        if (!this.panClient.tagExists(sgTag, this.tagXpath)) {

```

Comment:
hopefully not needed. We should assume if call succeeded, its created. Other wise its a PAN bug.